Add support for showing a spinner in an <input>

rr-codes opened the pull request:
#### 96729fcd8ccfe69a87bf1d779b992f6ecd7a6978
<pre>
Add support for showing a spinner in an &lt;input&gt;
<a href="https://bugs.webkit.org/show_bug.cgi?id=242067">https://bugs.webkit.org/show_bug.cgi?id=242067</a>
rdar://94641774

Reviewed by Devin Rousso.

This adds support for an enum case to show a loading indicator in an input element.

* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:]):
adds an accessibility text attribute to satisfy the new enum for the switch-case

* Source/WebCore/css/html.css:
(input::-webkit-loading-auto-fill-button):
(#endif):
this adds the loading indicator SVG to a new selector in the css file

* Source/WebCore/html/HTMLTextFormControlElement.h:
Adds the new enum case

* Source/WebCore/html/TextFieldInputType.cpp:
(WebCore::autoFillButtonTypeToAccessibilityLabel):
(WebCore::autoFillButtonTypeToAutoFillButtonText):
(WebCore::autoFillButtonTypeToAutoFillButtonPseudoClassName):
(WebCore::isAutoFillButtonTypeChanged):
Maps the enum to a `ShadowPseudoIds` attribute

* Source/WebCore/html/shadow/ShadowPseudoIds.cpp:
(WebCore::ShadowPseudoIds::webkitLoadingAutoFillButton):
* Source/WebCore/html/shadow/ShadowPseudoIds.h:
Adds a new function to get the CSS selector for the loading indicator

* Source/WebCore/platform/LocalizedStrings.cpp:
(WebCore::AXAutoFillLoadingLabel):
* Source/WebCore/platform/LocalizedStrings.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::toAutoFillButtonType):
(WebCore::toInternalsAutoFillButtonType):
* Source/WebCore/testing/Internals.h:
Adds a mapping from `Internals::AutoFillButtonType` to `AutoFillButtonType` for the Loading case

* Source/WebKit/WebProcess/InjectedBundle/API/Cocoa/WKWebProcessPlugInNodeHandle.mm:
(toAutoFillButtonType):
(toWKAutoFillButtonType):
Adds a mapping between `WebCore::AutoFillButtonType` and `_WKAutoFillButtonType` for the Loading case

* Source/WebKit/WebProcess/InjectedBundle/API/Cocoa/WKWebProcessPlugInNodeHandlePrivate.h:
Adds the new enum case

* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundleNodeHandle.cpp:
(toAutoFillButtonType):
(toWKAutoFillButtonType):
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundleNodeHandlePrivate.h:
Adds a mapping between `kWKAutoFillButtonTyp` and `WebCore::AutoFillButtonType` for the Loading case

Canonical link: <a href="https://commits.webkit.org/252146@main">https://commits.webkit.org/252146@main</a>
</pre>

dcrousso commented on Source/WebCore/css/html.css:
```diff
@@ -679,6 +679,21 @@ input::-webkit-credit-card-auto-fill-button:active {
 
+input::-webkit-loading-auto-fill-button {
+#if defined(HAVE_OS_DARK_MODE_SUPPORT) && HAVE_OS_DARK_MODE_SUPPORT
+@media (prefers-dark-interface) {
```
I don't think you can nest `@media` inside rulesets like this 😅

i'd move this down below into the existing grouping
```diff
 #if defined(HAVE_OS_DARK_MODE_SUPPORT) && HAVE_OS_DARK_MODE_SUPPORT
 @media (prefers-dark-interface) {
      input {
         background-color: transparent;
     }
+
+    input::-webkit-loading-auto-fill-button {
+        background-image: url(...);
+    }
 }
```
or just move it outside of this `input::-webkit-loading-auto-fill-button`

dcrousso commented on Source/WebCore/css/html.css:
```diff
@@ -679,6 +679,21 @@ input::-webkit-credit-card-auto-fill-button:active {
+    background-image: url('data:image/svg+xml,<svg width="22" height="12" viewBox="0 0 32 32" xmlns="http://www.w3.org/2000/svg" xml:space="preserve"><path fill="rgb(55,55,55)" d="M19.36,12.64L19.36,12.64c-0.88-0.88-0.88-2.3,0-3.18l4.6-4.6c0.88-0.88,2.3-0.88,3.18,0l0,0c0.88,0.88,0.88,2.3,0,3.18l-4.6,4.6C21.66,13.52,20.24,13.52,19.36,12.64z"><animate attributeName="fill" begin="0s" dur="1s" values ...
+    width: 22px;
+    height: 12px;
+    margin-left: 3px;
+    margin-right: 2px;
```
do we need to do anything special here for RTL?  maybe use `margin-inline-start` and `margin-inline-end` instead?

rr-codes replied:
I just followed the same convention that `input::-webkit-credit-card-auto-fill-button` was doing

dcrousso commented on Source/WebCore/testing/Internals.h:
```diff
@@ -337,7 +337,7 @@ class Internals final : public RefCounted<Internals>, private ContextDestruction
     void setAutoFilledAndViewable(HTMLInputElement&, bool enabled);
     void setAutoFilledAndObscured(HTMLInputElement&, bool enabled);
-    enum class AutoFillButtonType { None, Contacts, Credentials, StrongPassword, CreditCard };
+    enum class AutoFillButtonType { None, Contacts, Credentials, StrongPassword, CreditCard, Loading };
```
I think you also want to add this to `Source/WebCore/testing/Internals.idl` and write some tests :)

rr-codes replied:
yup, Aditya is helping me do that now

dcrousso commented on Source/WebCore/css/html.css:
```diff
@@ -679,6 +679,21 @@ input::-webkit-credit-card-auto-fill-button:active {
+    background-image: url('data:image/svg+xml,<svg width="22" height="12" viewBox="0 0 32 32" xmlns="http://www.w3.org/2000/svg" xml:space="preserve"><path fill="rgb(55,55,55)" d="M19.36,12.64L19.36,12.64c-0.88-0.88-0.88-2.3,0-3.18l4.6-4.6c0.88-0.88,2.3-0.88,3.18,0l0,0c0.88,0.88,0.88,2.3,0,3.18l-4.6,4.6C21.66,13.52,20.24,13.52,19.36,12.64z"><animate attributeName="fill" begin="0s" dur="1s" values ...
+}
+#endif
+    background-image: url('data:image/svg+xml,<svg width="22" height="12" viewBox="0 0 32 32" xmlns="http://www.w3.org/2000/svg" xml:space="preserve"><path fill="rgb(55,55,55)" d="M19.36,12.64L19.36,12.64c-0.88-0.88-0.88-2.3,0-3.18l4.6-4.6c0.88-0.88,2.3-0.88,3.18,0l0,0c0.88,0.88,0.88,2.3,0,3.18l-4.6,4.6C21.66,13.52,20.24,13.52,19.36,12.64z"><animate attributeName="fill" begin="0s" dur="1s" values ...
```
Is there a reason to use this instead of `mask-image`?  I see that both are used in this file, but not sure why one over the other 🤔

rr-codes replied:
there is; Wenson and I discovered that `mask-image` doesn't work with SVGs that are animated apparently

nt1m replied:
please file a bug

dcrousso commented on Source/WebCore/css/html.css:
```diff
@@ -433,6 +433,16 @@ input {
+        margin-left: 3px;
+        margin-right: 2px;
+        flex: none;
+        user-select: none;
```
you don't have to repeat all of these since the `input::-webkit-loading-auto-fill-button` will still apply

only put stuff here that you want to override from the `input::-webkit-loading-auto-fill-button` below

dcrousso commented on LayoutTests/fast/forms/auto-fill-button/input-loading-auto-fill-button.html:
```diff
@@ -0,0 +1,16 @@
+if (window.internals) {
+    const input = document.querySelector("input");
+    window.internals.setShowAutoFillButton(input, "Loading");
+}
```
NIT: you could rewrite this as `window.internals?.setShowAutoFillButton(input, "Loading")`

rr-codes replied:
oh, null coalescing is in JS too? I thought it was just a TypeScript thing. I'll change it

dcrousso replied:
this is optional chaining actually 🙂

both optional chaining and nullish coalescing are JS features, not TypeScript features 😉

dcrousso commented on LayoutTests/fast/forms/auto-fill-button/input-loading-auto-fill-button.html:
```diff
@@ -0,0 +1,16 @@
+<html>
+<body>
+<p>This tests that the Loading AutoFill button renders. It can only be tested in the test tool.</p>
+<div id="container">
```
NIT: Is this necessary, or can we just have the `<input type="text">` be toplevel?

rr-codes replied:
I'm not sure if its necessary; I just followed the convention of the other autofill test HTML files

pxlcoder replied:
I think you can remove it.

rr-codes committed:
Add support for showing a spinner in an <input>
https://bugs.webkit.org/show_bug.cgi?id=242067
rdar://94641774

Reviewed by Devin Rousso.

This adds support for an enum case to show a loading indicator in an input element.

* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:]):
adds an accessibility text attribute to satisfy the new enum for the switch-case

* Source/WebCore/css/html.css:
(input::-webkit-loading-auto-fill-button):
(#endif):
this adds the loading indicator SVG to a new selector in the css file

* Source/WebCore/html/HTMLTextFormControlElement.h:
Adds the new enum case

* Source/WebCore/html/TextFieldInputType.cpp:
(WebCore::autoFillButtonTypeToAccessibilityLabel):
(WebCore::autoFillButtonTypeToAutoFillButtonText):
(WebCore::autoFillButtonTypeToAutoFillButtonPseudoClassName):
(WebCore::isAutoFillButtonTypeChanged):
Maps the enum to a `ShadowPseudoIds` attribute

* Source/WebCore/html/shadow/ShadowPseudoIds.cpp:
(WebCore::ShadowPseudoIds::webkitLoadingAutoFillButton):
* Source/WebCore/html/shadow/ShadowPseudoIds.h:
Adds a new function to get the CSS selector for the loading indicator

* Source/WebCore/platform/LocalizedStrings.cpp:
(WebCore::AXAutoFillLoadingLabel):
* Source/WebCore/platform/LocalizedStrings.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::toAutoFillButtonType):
(WebCore::toInternalsAutoFillButtonType):
* Source/WebCore/testing/Internals.h:
Adds a mapping from `Internals::AutoFillButtonType` to `AutoFillButtonType` for the Loading case

* Source/WebKit/WebProcess/InjectedBundle/API/Cocoa/WKWebProcessPlugInNodeHandle.mm:
(toAutoFillButtonType):
(toWKAutoFillButtonType):
Adds a mapping between `WebCore::AutoFillButtonType` and `_WKAutoFillButtonType` for the Loading case

* Source/WebKit/WebProcess/InjectedBundle/API/Cocoa/WKWebProcessPlugInNodeHandlePrivate.h:
Adds the new enum case

* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundleNodeHandle.cpp:
(toAutoFillButtonType):
(toWKAutoFillButtonType):
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundleNodeHandlePrivate.h:
Adds a mapping between `kWKAutoFillButtonTyp` and `WebCore::AutoFillButtonType` for the Loading case

Canonical link: https://commits.webkit.org/252146@main
